File: 07-interfaces/python/animal.py
"""
animal.py - CLASE ABSTRACTA (la jerarquia)

Esto es lo que un animal ES. Tiene ESTADO compartido (nombre, edad)
y comportamiento comun (presentarse). Por eso es una clase abstracta
y no una interfaz.
"""

from abc import ABC, abstractmethod


class Animal(ABC):

    def __init__(self, nombre, edad):
        self._nombre = nombre
        self._edad = edad

    @abstractmethod
    def sonido(self):
        """Cada especie hace su ruido."""

    @abstractmethod
    def get_grupo(self):
        """Cada rama de la jerarquia dice a que grupo pertenece."""

    def get_nombre(self):
        return self._nombre

    def presentarse(self):
        # type(self).__name__ es el equivalente de getClass().getSimpleName()
        print(f"{type(self).__name__} {self._nombre} "
              f"es un {self.get_grupo()}, dice: {self.sonido()}")

File: 07-interfaces/java/Animal.java
/*
 * Animal.java - CLASE ABSTRACTA (la jerarquia)
 *
 * Esto es lo que un animal ES. Tiene ESTADO compartido (nombre, edad)
 * y comportamiento comun (presentarse). Por eso es una clase abstracta
 * y no una interfaz.
 */
public abstract class Animal {

    protected String nombre;
    protected int edad;

    public Animal(String nombre, int edad) {
        this.nombre = nombre;
        this.edad = edad;
    }

    // Cada especie hace su ruido
    public abstract String sonido();

    // Cada rama de la jerarquia dice a que grupo pertenece
    public abstract String getGrupo();

    public String getNombre() {
        return this.nombre;
    }

    public void presentarse() {
        System.out.println(getClass().getSimpleName() + " " + nombre
                           + " es un " + getGrupo() + ", dice: " + sonido());
    }
}
